nova/scheduler/filters/shard_filter.py

```diff
@@ -82,9 +84,19 @@ def _update_cache(self):
             data = resp.json()
             for project in data['projects']:
                 project_id = project['id']
-                shards = [t for t in project['tags']
-                          if t.startswith(self._SHARD_PREFIX)]
+                tags = project['tags']
+                shards = [t for t in tags if t.startswith(self._SHARD_PREFIX)]
+                # NOTE(jakob): If the project has more than one shard enabled
+                # in _any_ AZ, consider all shards is enabled in _all_ AZs.
+                # The proper `sharding_enabled` tag is appended if not already
+                # in the actual list of tags.
+                shard_counts = Counter("".join(shard.split("-")[:2])
+                                       for shard in shards)
+                all_shards = any(v > 1 for v in shard_counts.values())
                 self._PROJECT_SHARD_CACHE[project_id] = shards
+                if all_shards and not self._ALL_SHARDS in tags:
+                    self._PROJECT_SHARD_CACHE[project_id].append(
+                        self._ALL_SHARDS)
```

Comment:
Could you explain, why you did that? I thought the goal was to short-circuit the ShardFilter, if the `sharding_enabled` tag is in there. Why do we change the meaning of setting a list of shards to be enabled?

If we want currently-multishard projects to get all shards, we can run a script to add the `sharding_enabled` tag.

Reply:
Yeah, that makes more sense. I thought that this will make a 3/4 situation (`a-0, a-1, b-0`) behave correctly in the sense that we wouldn't allow it and want it to mean _all_ shards (in the example `b-1` as well).

But you're right, we can do this better through scripting, and save the extra code here.